File: 2024/07/solve.hs
#!/usr/bin/env runhaskell
{-# LANGUAGE BangPatterns #-}

import Control.Monad (replicateM)
import Data.List

lineParse :: String -> (Int, [Int])
lineParse s = case words s of
  (x : xs) -> (read (init x), read <$> xs)
  _ -> error "Invalid input"

concatenate :: Int -> Int -> Int
concatenate x y = x * (10 ^ length (show y)) + y

calibrationResult :: [Int -> Int -> Int] -> [(Int, [Int])] -> Int
calibrationResult ops dat = sum $ fst <$> filter checkTarget dat
  where
    checkTarget (target, nums) = elem target $ evaluate nums <$> replicateM (length nums - 1) ops
    evaluate (x : xs) operatorList = foldl' (\acc (op, n) -> op acc n) x (zip operatorList xs)
    evaluate [] _ = error "Empty number list"

main :: IO ()
main = do
  input <- readFile "input"
  let dat = lineParse <$> lines input
      !partOne = calibrationResult [(+), (*)] dat
      !partTwo = calibrationResult [(+), (*), concatenate] dat
  print partOne
  print partTwo

File: 2024/07/solve.py
#!/usr/bin/env python3
from itertools import product
from operator import mul, add
from functools import partial, reduce

def calibration_result(data, ops):
    return sum(
        x
        for x, ns in data
        if any(
            val == x
            for val in (
                reduce(lambda acc, opn: partial(*opn)(acc), zip(combo, ns[1:]), ns[0])
                for combo in product(ops, repeat=len(ns) - 1)
            )
        )
    )

data = [ (int(x), tuple(map(int, ns.split()))) for x, ns in (line.split(":") for line in open("input")) ]

print(calibration_result(data, (add, mul)))  # part 1
print(calibration_result(data, (add, mul, lambda x, y: int(str(y)+str(x)))))  # part 2
